File: js-impl/julia.js
export function generateJulia(width, height, real, imaginary) {
    const scaleX = 3.0 / width;
    const scaleY = 3.0 / height;
    const paramI = 1.5;
    const paramR = 1.5;

    const data = new Uint8ClampedArray(width * height);

    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++) {
        let zx = x * scaleX - paramR;
        let zy = y * scaleY - paramI;
        let iterIndex = getIterIndex(zx, zy, real, imaginary);
        data[y * width + x] = iterIndex;
      }
    }

    return data;
  }

  function getIterIndex(x, y, real, imaginary) {
    let zx = x;
    let zy = y;
    let iterIndex = 0;

    while (iterIndex < 255) {
      const xtemp = zx * zx - zy * zy + real;
      zy = 2 * zx * zy + imaginary;
      zx = xtemp;
      if (zx * zx + zy * zy > 4) {
        // Pythagoras theorem in complex plane
        break;
      }
      iterIndex++;
    }

    return iterIndex;
  }

File: cpp-impl/julia.cc
#include <iostream>
#include <string>
#include <complex>
#include <vector>

#include "julia.h"
#include "emscripten/bind.h"

using namespace emscripten;
using namespace std;


Julia::Julia() {}

std::vector<int> Julia::generateJuliaSet(unsigned int width, unsigned int height, float real, float imaginary) {
    // c = imaginary * i + real

    float scale_x = 3.0 / width;
    float scale_y = 3.0 / height;
    // 2.0 border + 2
    // int* data = new int[width * height];
    std::vector<int> data;
    data.resize(width * height);
    complex<float> c(real, imaginary);
    // initial value of z
    float param_i = 1.5;
    float param_r = 1.5;

    for (unsigned int y = 0; y < height; ++y) {
        for (unsigned int x = 0; x < width; ++x) {
            complex<float> z(x * scale_x - param_r, y * scale_y - param_i);
            unsigned int iter_index = Julia::get_iter_index(z, c);
            data[y * width + x] = static_cast<int>(iter_index);
        }
    }

    return data;
}

unsigned int Julia::get_iter_index(complex<float> z, complex<float> c) {
    unsigned int iter_index = 0;
    while (iter_index < 255) {
        if (norm(z) > 2.0) {
            break;
        }
        z = z * z + c;
        iter_index++;
    }
    return iter_index;

}

EMSCRIPTEN_BINDINGS (c) {
  register_vector<int>("VectorInt");

  class_<Julia>("Julia")
    .constructor()
    .function("generateJuliaSet", &Julia::generateJuliaSet);
}
